src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,215 @@
+						}
+					}
+				}
+
+				if (targetPaths.length === 0) {
+					this.logger.trace(`No matching cpd ${language} target files found. Nothing to execute.`);
+					return [];
+				}
+				this.logger.trace(`About to run CPD (${language}). Targets: ${targetPaths.length}`);
+
+				const selectedTargets = targetPaths.join(',');
+				const stdout = await CpdWrapper.execute(selectedTargets, language, await this.config.getMinimumTokens(ENGINE.CPD));
+				results.push(...this.processStdOut(stdout));
+				
+				this.logger.trace(`Found ${results.length} for CPD`);
+			}   
+			return results;
+		} catch (e) {
+			this.logger.trace('Cpd evaluation failed: ' + (e.message || e));
+			throw new SfdxError(e.message || e);
+		}
+	}
+
+	private validateFileType(language: string, fileName: string): boolean {
+		if (this.languageToPathMap.has(language)){
+			for (const fileType of this.languageToPathMap.get(language)) {
+				if (fileName.endsWith(fileType)) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	protected processStdOut(stdout: string): RuleResult[] {
+		let violations: RuleResult[] = [];
+
+		this.logger.trace(`Output received from CPD: ${stdout}`);
+
+		const cpdEnd = '</pmd-cpd>';
+		const xmlStart = stdout.indexOf('<?xml');
+		const xmlEnd = stdout.lastIndexOf(cpdEnd);
+		if (xmlStart != -1 && xmlEnd != -1) {
+			const cpdXml = stdout.slice(xmlStart, xmlEnd + cpdEnd.length);
+			const cpdJson = xml2js(cpdXml, {compact: false, ignoreDeclaration: true}); 
+
+			const duplications =  cpdJson.elements[0].elements;
+			if (duplications) {
+				violations = this.jsonToRuleResults(duplications);
+			}
+		}
+
+		if (violations.length > 0) {
+			this.logger.trace('Found rule violations.');
+		}
+
+		return violations;
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	private jsonToRuleResults(duplications: any): RuleResult[] {    
+		const ruleResults: RuleResult[] = [];
+
+		for (const duplication of duplications) {
+		
+			const parts = duplication.elements;
+			const codeFragment = parts.find(part => {return part.name === 'codefragment'}).elements[0].cdata;
+			const occurences = parts.filter(part => {return part.name === 'file'});
+			const codeFragmentID = checksum(codeFragment).slice(0,7);
+			// TODO: check for collisions (cpd can return multiple duplications for the same code fragment, so these have the same checksum currently)
+
+			let occCount = 1;
+			for (const occ of occurences) {
+				// create a violation for each occurence of the code fragment
+				const violation: RuleViolation = {
+					line: occ.attributes.line,
+					column: occ.attributes.column,
+					endLine: occ.attributes.endline,
+					endColumn: occ.attributes.endcolumn,
+					ruleName: this.cpdCatalog.rules[0].name,
+					severity: 3,
+					message: `${codeFragmentID}: ${occCount} of ${occurences.length} duplication segments detected. ${duplication.attributes.lines} line(s), ${duplication.attributes.tokens} tokens.`, 
```

Comment:
Can we have this message sourced from a js file so that we can later get this reviewed by a doc writer?